### ISSUE TITLE
Question

### ISSUE THREAD

**Issue by sanskriti-s** [open-policy-agent/opa#2438]
How do you create a mutating webhook that can create a list in the spec in a deployment and add a value to it?

# Original deployment
`kubectl get deployment -n default nginx-deployment -o json | jq '.spec'`:
```
{
  "progressDeadlineSeconds": 600,
  "replicas": 1,
  "revisionHistoryLimit": 10,
  "selector": {
    "matchLabels": {
      "app": "nginx"
    }
  },
  "strategy": {
    "rollingUpdate": {
      "maxSurge": "25%",
      "maxUnavailable": "25%"
    },
    "type": "RollingUpdate"
  },
  "template": {
    "metadata": {
      "creationTimestamp": null,
      "labels": {
        "app": "nginx"
      }
    },
    "spec": {
      "containers": [
        {
          "image": "nginx:1.14.1",
          "imagePullPolicy": "IfNotPresent",
          "name": "nginx",
          "ports": [
            {
              "containerPort": 80,
              "protocol": "TCP"
            }
          ],
          "resources": {},
          "terminationMessagePath": "/dev/termination-log",
          "terminationMessagePolicy": "File"
        }
      ],
      "dnsPolicy": "ClusterFirst",
      "restartPolicy": "Always",
      "schedulerName": "default-scheduler",
      "securityContext": {},
      "terminationGracePeriodSeconds": 30
    }
  }
}
```

# Wanted Behavior
I'm trying to make a mutating webhook that mutates a deployment by adding the command so that
`kubectl get deployment -n default nginx-deployment -o json | jq '.spec'`:
```
  "progressDeadlineSeconds": 600,
  "replicas": 1,
  "revisionHistoryLimit": 10,
  "selector": {
    "matchLabels": {
      "app": "nginx"
    }
  },
  "strategy": {
    "rollingUpdate": {
      "maxSurge": "25%",
      "maxUnavailable": "25%"
    },
    "type": "RollingUpdate"
  },
  "template": {
    "metadata": {
      "creationTimestamp": null,
      "labels": {
        "app": "nginx"
      }
    },
    "spec": {
      "containers": [
        {
          "image": "nginx:1.14.1",
          "imagePullPolicy": "IfNotPresent",
          "name": "nginx",
          "ports": [
            {
              "containerPort": 80,
              "protocol": "TCP"
            }
            "command": [
             - test1
             ]
          ],
          "resources": {},
          "terminationMessagePath": "/dev/termination-log",
          "terminationMessagePolicy": "File"
        }
      ],
      "dnsPolicy": "ClusterFirst",
      "restartPolicy": "Always",
      "schedulerName": "default-scheduler",
      "securityContext": {},
      "terminationGracePeriodSeconds": 30
    }
  }
}
```

# Actual Behavior
No changes to original deployment

# Rego template
```
package library.kubernetes.admission.mutating
 
###########################################################################
# Implementation of the k8s admission control external webhook interface,
# combining validating and mutating admission controllers
###########################################################################
 
main = {
    "apiVersion": "admission.k8s.io/v1beta1",
    "kind": "AdmissionReview",
    "response": response,
}
 
default response = {"allowed": true}
 
# non-patch response i.e. validation response
response = x {
    count(patch) == 0
 
    x := {
        "allowed": false,
        "status": {"reason": reason},
    }
 
    reason = concat(", ", deny)
    reason != ""
}
 
# patch response i.e. mutating respone
response = x {
    count(patch) > 0
 
    # if there are missing leaves e.g. trying to add a label to something that doesn't
    # yet have any, we need to create the leaf nodes as well
 
    fullPatches := ensureParentPathsExist(cast_array(patch))
 
    x := {
        "allowed": true,
        "patchType": "JSONPatch",
        "patch": base64.encode(json.marshal(fullPatches)),
    }
}
 
isValidRequest {
    # not sure if this might be a race condition, it might get called before
    # all the validation rules have been run
    count(deny) == 0
}
 
isCreateOrUpdate {
    isCreate
}
 
isCreateOrUpdate {
    isUpdate
}
 
isCreate {
    input.request.operation == "CREATE"
}
 
isUpdate {
    input.request.operation == "UPDATE"
}
 
###########################################################################
# PATCH helpers
# Note: These rules assume that the input is an object
# not an AdmissionRequest, because labels and annotations
# can apply to various sub-objects within a request
# So from the context of an AdmissionRequest they need to
# be called like
#   hasLabelValue("foo", "bar") with input as input.request.object
# or
#   hasLabelValue("foo", "bar") with input as input.request.oldObject
###########################################################################
 
hasLabels(obj) {
    obj.metadata.labels
}
 
hasLabel(obj, label) {
    obj.metadata.labels[label]
}
 
hasLabelValue(obj, key, val) {
    obj.metadata.labels[key] = val
}
 
hasAnnotations(obj) {
    obj.metadata.annotations
}
 
hasAnnotation(obj, annotation) {
    obj.metadata.annotations[annotation]
}
 
hasAnnotationValue(obj, key, val) {
    obj.metadata.annotations[key] = val
}
 
###########################################################################
# makeLabelPatch creates a label patch
# Labels can exist on numerous child objects e.g. Deployment.template.metadata
# Use pathPrefix to specify a lower level object, or pass "" to select the
# top level object
# Note: pathPrefix should have a leading '/' but no trailing '/'
###########################################################################
 
makeSpecPatch(op, key, value, pathPrefix) = patchCode {
    patchCode = {
        "op": op,
        "path": concat("/", [pathPrefix, "spec/template/spec/containers/0/", replace(key, "/", "~1")]),
        "value": value,
    }
}

makeLabelPatch(op, key, value, pathPrefix) = patchCode {
    patchCode = {
        "op": op,
        "path": concat("/", [pathPrefix, "metadata/labels", replace(key, "/", "~1")]),
        "value": value,
    }
}

makeAnnotationPatch(op, key, value, pathPrefix) = patchCode {
    patchCode = {
        "op": op,
        "path": concat("/", [pathPrefix, "metadata/annotations", replace(key, "/", "~1")]),
        "value": value,
    }
}
 
# Given array of JSON patches create and prepend new patches that create missing paths.
ensureParentPathsExist(patches) = result {
    # Convert patches to a set
    paths := {p.path | p := patches[_]}
 
    # Compute all missing subpaths.
    #    Iterate over all paths and over all subpaths
    #    If subpath doesn't exist, add it to the set after making it a string
    missingPaths := {sprintf("/%s", [concat("/", prefixPath)]) |
        paths[path]
        pathArray := split(path, "/")
        pathArray[i] # walk over path
        i > 0 # skip initial element
 
        # array of all elements in path up to i
        prefixPath := [pathArray[j] | pathArray[j]; j < i; j > 0] # j > 0: skip initial element
        walkPath := [toWalkElement(x) | x := prefixPath[_]]
        not inputPathExists(walkPath) with input as input.request.object
    }
 
    # Sort paths, to ensure they apply in correct order
    ordered_paths := sort(missingPaths)
 
    # Return new patches prepended to original patches.
    #  Don't forget to prepend all paths with a /
    new_patches := [{"op": "add", "path": p, "value": {}} |
        p := ordered_paths[_]
    ]
 
    result := array.concat(new_patches, patches)
}
 
# Check that the given @path exists as part of the input object.
inputPathExists(path) {
    walk(input, [path, _])
}
 
toWalkElement(str) = str {
    not re_match("^[0-9]+$", str)
}
 
toWalkElement(str) = x {
    re_match("^[0-9]+$", str)
    x := to_number(str)
}

############################################################
# DENY rules
############################################################
 
# Don't allow container images with no version tag, or a version tag that doesn't contain at least one digit
 
missingImageVersion(imageName) {
    not re_match(`.*:.*\d.*$`, imageName)
}
 
deny[msg] {
    input.request.kind.kind = "Deployment"
    badImages = {image |
        image = input.request.object.spec.template.spec.containers[_].image
        missingImageVersion(image, true)
    }
 
    count(badImages) > 0
    names = concat(", ", badImages)
    msg = sprintf("Container images must specify a version (%s)", [names])
}
 
############################################################
# PATCH rules
#
# Note: All patch rules should start with `isValidRequest` and `isCreateOrUpdate`
############################################################
 
patch[patchCode] {
    isValidRequest
    isCreateOrUpdate
    input.request.kind.kind == "Deployment"
    hasAnnotationValue(input.request.object, "test-mutation", "true")
    patchCode = makeSpecPatch("add", "command/0", "test1", "")
}
```
I'm able to make other mutations so the issue is most likely in how I'm trying to create a list.
Thanks in advance,
Sanskriti

**Comment by patrick-east** [comment]
I'd highly recommend troubleshooting this by writing unit tests for your policies: https://www.openpolicyagent.org/docs/latest/policy-testing/

If you know what a valid working patch looks like you can then work towards writing the policy to dynamically generate it.

Assuming you had a deployment with the right annotations it would generate a patch like:

```
        {
            "op": "add",
            "path": "/spec/template/spec/containers/0//command~10",
            "value": "test1"
        }
```

Which I guess then becomes a question of whether or not that is a valid path specification for the API server.. You'll have to consult with the JSON patch specification, at a glance I'm not sure if thats the right thing for what you want.

The other thing to check is if there are any errors in the API server logs if it isn't able to apply the patch.

**Comment by sanskriti-s** [comment]
I tried putting my actual input in the playground:
```
{
    "apiVersion": "admission.k8s.io/v1beta1",
    "kind": "AdmissionReview",
    "request": {
        "dryRun": false,
        "kind": {
            "group": "apps",
            "kind": "Deployment",
            "version": "v1"
        },
        "name": "nginx-deployment",
        "namespace": "default",
        "object": {
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "metadata": {
                "creationTimestamp": null,
                "labels": {
                    "app": "nginx"
                },
                "name": "nginx-deployment",
                "namespace": "default"
            },
            "spec": {
                "progressDeadlineSeconds": 600,
                "replicas": 1,
                "revisionHistoryLimit": 10,
                "selector": {
                    "matchLabels": {
                        "app": "nginx"
                    }
                },
                "strategy": {
                    "rollingUpdate": {
                        "maxSurge": "25%",
                        "maxUnavailable": "25%"
                    },
                    "type": "RollingUpdate"
                },
                "template": {
                    "metadata": {
                        "creationTimestamp": null,
                        "labels": {
                            "app": "nginx"
                        }
                    },
                    "spec": {
                        "containers": [
                            {
                                "image": "nginx:1.14.2",
                                "imagePullPolicy": "IfNotPresent",
                                "name": "nginx",
                                "ports": [
                                    {
                                        "containerPort": 80,
                                        "protocol": "TCP"
                                    }
                                ],
                                "resources": {},
                                "terminationMessagePath": "/dev/termination-log",
                                "terminationMessagePolicy": "File"
                            }
                        ],
                        "dnsPolicy": "ClusterFirst",
                        "restartPolicy": "Always",
                        "schedulerName": "default-scheduler",
                        "securityContext": {},
                        "terminationGracePeriodSeconds": 30
                    }
                }
            },
            "status": {}
        },
        "oldObject": null,
        "operation": "CREATE",
        "options": {
            "apiVersion": "meta.k8s.io/v1",
            "kind": "CreateOptions"
        },
        "requestKind": {
            "group": "apps",
            "kind": "Deployment",
            "version": "v1"
        },
        "requestResource": {
            "group": "apps",
            "resource": "deployments",
            "version": "v1"
        },
        "resource": {
            "group": "apps",
            "resource": "deployments",
            "version": "v1"
        },
        "uid": "735a3b6e-ff36-43f3-b25d-f32706e808a0",
        "userInfo": {
            "groups": [
                "system:masters",
                "system:authenticated"
            ],
            "username": "kubernetes-admin"
        }
    }
}
```
And I'm getting an empty patch as the output:
```
{
    "deny": [],
    "isCreate": true,
    "isCreateOrUpdate": true,
    "isValidRequest": true,
    "main": {
        "apiVersion": "admission.k8s.io/v1beta1",
        "kind": "AdmissionReview",
        "response": {
            "allowed": true
        }
    },
    "patch": [],
    "response": {
        "allowed": true
    }
}
```

Do you know what could be causing this?

**Comment by patrick-east** [comment]
I'm guessing you didn't write the unit tests for it... let me help: https://gist.github.com/patrick-east/01c67afcab9aaf59c4603fb2d41355fb

If you test them, we can see the failures pretty quickly in the output:

```
{11:01} ✘ /t/i/2438 ❯ opa test -v ./policy.rego ./policy_test.rego
FAILURES
--------------------------------------------------------------------------------
data.test.library.kubernetes.admission.mutating.test_mutating_cmd: FAIL (941.033µs)

  query:1                    Enter data.test.library.kubernetes.admission.mutating.test_mutating_cmd = _
  ./policy_test.rego:3       | Enter data.test.library.kubernetes.admission.mutating.test_mutating_cmd
  ./policy.rego:209          | | Enter data.library.kubernetes.admission.mutating.patch
  ./policy.rego:45           | | | Enter data.library.kubernetes.admission.mutating.isValidRequest
  ./policy.rego:191          | | | | Enter data.library.kubernetes.admission.mutating.deny
  ./policy.rego:194          | | | | | Enter image = input.request.object.spec.template.spec.containers[_].image; data.library.kubernetes.admission.mutating.missingImageVersion(image, true)
  ./policy.rego:187          | | | | | | Enter data.library.kubernetes.admission.mutating.missingImageVersion
  ./policy.rego:188          | | | | | | | Fail not re_match(".*:.*\\d.*$", imageName)
  ./policy.rego:195          | | | | | | Fail data.library.kubernetes.admission.mutating.missingImageVersion(image, true)
  ./policy.rego:198          | | | | | Fail gt(__local68__, 0)
  ./policy.rego:99           | | | Enter data.library.kubernetes.admission.mutating.hasAnnotationValue
  ./policy.rego:100          | | | | Fail obj.metadata.annotations[key] = val
  ./policy.rego:213          | | | Fail data.library.kubernetes.admission.mutating.hasAnnotationValue(__local83__, "test-mutation", "true")
  ./policy.rego:55           | | | Enter data.library.kubernetes.admission.mutating.isCreateOrUpdate
  ./policy.rego:56           | | | | Fail data.library.kubernetes.admission.mutating.isUpdate
  ./policy_test.rego:109     | | Fail patches[expected]
  query:1                    | Fail data.test.library.kubernetes.admission.mutating.test_mutating_cmd = _

SUMMARY
--------------------------------------------------------------------------------
data.test.library.kubernetes.admission.mutating.test_mutating_cmd: FAIL (941.033µs)
--------------------------------------------------------------------------------
FAIL: 1/1
```

Note the trace events that have a `Fail`. Those are expressions that were either false or undefined. Checking on each one in the policy the one that stands out the most is..

```
./policy.rego:213          | | | Fail data.library.kubernetes.admission.mutating.hasAnnotationValue(__local83__, "test-mutation", "true")
```

If I then look at that rule, and the input, I see that yes indeed there is a missing `"test-mutation": "true"` annotation on the input. Note, I don't know if that is a wrong path or it was just missing on the original manifest.

Correcting that for the test by just adding in 

```
"annotations": {"test-mutation": "true"},
```
To the test input

We now get success:

```
{11:09} /t/i/2438 ❯ opa test -v ./policy.rego ./policy_test.rego
data.test.library.kubernetes.admission.mutating.test_mutating_cmd: PASS (1.270693ms)
--------------------------------------------------------------------------------
PASS: 1/1
```
Note: the gist does not include this fix..

Hopefully that helps to unblock you.

**Comment by patrick-east** [comment]
Going to close this out for now. If any more issues come up feel free to reopen or file a new issue.
